File: server/dive_utils/model_pack.py
"""Portable model-pack layout and safe, streaming ZIP extraction."""

from pathlib import Path, PurePosixPath
import re
import shutil
import stat
from zipfile import ZipFile

MAX_FILES = 100000
MAX_BYTES = 100 * 1024**3


def model_pack_paths(names):
    """Map ZIP members to paths relative to the pipeline directory."""
    clean = []
    for name in names:
        path = name.replace('\\', '/')
        parts = path.rstrip('/').split('/')
        if (
            path.startswith('/')
            or any(p in ('..', '') or ':' in p for p in parts)
            or '\x00' in path
        ):
            raise ValueError(f'Unsafe model archive path: {name}')
        if '__MACOSX' in parts or parts[-1] == '.DS_Store':
            continue
        clean.append((name, str(PurePosixPath(path))))

    # Standard VIAME layouts, flat packs, and any of these inside one wrapper.
    candidates = ['configs/pipelines/', 'pipelines/', '']
    wrappers = sorted({p.split('/')[0] for _, p in clean if '/' in p})
    for wrapper in wrappers:
        candidates.extend([f'{wrapper}/configs/pipelines/', f'{wrapper}/pipelines/', f'{wrapper}/'])
    root = next(
        (
            prefix
            for prefix in candidates
            if any(
                p.startswith(prefix) and '/' not in p[len(prefix) :] and p.endswith('.pipe')
                for _, p in clean
            )
        ),
        None,
    )
    if root is None:
        raise ValueError(
            'The ZIP must contain at least one .pipe file in a supported model layout.'
        )
    result = {}
    seen = set()
    for original, path in clean:
        if not path.startswith(root):
            continue
        relative = path[len(root) :]
        if not relative:
            continue
        key = relative.casefold()
        if key in seen:
            raise ValueError(f'Duplicate model archive path: {relative}')
        seen.add(key)
        result[original] = relative
    # A file must not also be a parent directory of another file.
    for relative in result.values():
        if any(str(p).casefold() in seen for p in PurePosixPath(relative).parents if str(p) != '.'):
            raise ValueError(f'Conflicting model archive path: {relative}')
    return result


def model_pack_name(filename):
    """Derive a portable directory name from an uploaded archive name."""
    name = re.sub(r'[^\w .-]', '_', Path(filename.replace('\\', '/')).stem).strip(' .')
    return name or 'Imported model'


def extract_model_pack(archive, destination):
    """Extract into a new staging directory, preserving model-relative paths."""
    with ZipFile(archive) as source:
        entries = source.infolist()
        if len(entries) > MAX_FILES or sum(e.file_size for e in entries) > MAX_BYTES:
            raise ValueError('Model archive exceeds the file count or uncompressed size limit.')
        for entry in entries:
            mode = entry.external_attr >> 16
            if stat.S_ISLNK(mode) or (stat.S_IFMT(mode) not in (0, stat.S_IFREG, stat.S_IFDIR)):
                raise ValueError('Model archives cannot contain symbolic links or special files.')
        files = [entry for entry in entries if not entry.is_dir()]
        mapping = model_pack_paths([entry.filename for entry in files])
        for entry in files:
            if entry.filename not in mapping:
                continue
            target = Path(destination) / mapping[entry.filename]
            target.parent.mkdir(parents=True, exist_ok=True)
            with source.open(entry) as src, target.open('xb') as dst:
                shutil.copyfileobj(src, dst)
        return mapping

File: client/platform/desktop/backend/native/modelPack.ts
/* Sequential streaming bounds memory and open files for large model packs. */
/* eslint-disable no-await-in-loop, no-restricted-syntax */
import path from 'path';
import fs from 'fs-extra';
import { pipeline as streamPipeline } from 'stream/promises';
import archiver from 'archiver';
import * as yauzl from 'yauzl';
import type { Pipe } from 'dive-common/apispec';
import { PipelinesFolderName, Settings } from 'platform/desktop/constants';

const MAX_FILES = 100000;
const MAX_BYTES = 100 * 1024 ** 3;

/** Keep relative model references intact while removing known archive wrappers. */
export function modelPackPaths(names: string[]): Map<string, string> {
  const clean = names.flatMap((name) => {
    const normalized = name.replace(/\\/g, '/');
    const parts = normalized.replace(/\/$/, '').split('/');
    if (normalized.startsWith('/') || normalized.includes('\0')
        || parts.some((part) => part === '..' || part === '' || part.includes(':'))) {
      throw new Error(`Unsafe model archive path: ${name}`);
    }
    if (parts.includes('__MACOSX') || parts[parts.length - 1] === '.DS_Store') return [];
    return [[name, path.posix.normalize(normalized)]];
  });
  const candidates = ['configs/pipelines/', 'pipelines/', ''];
  const wrappers = [...new Set(clean.filter(([, name]) => name.includes('/')).map(([, name]) => name.split('/')[0]))].sort();
  wrappers.forEach((wrapper) => candidates.push(`${wrapper}/configs/pipelines/`, `${wrapper}/pipelines/`, `${wrapper}/`));
  const root = candidates.find((prefix) => clean.some(([, name]) => name.startsWith(prefix)
    && !name.slice(prefix.length).includes('/') && name.endsWith('.pipe')));
  if (root === undefined) throw new Error('The ZIP must contain at least one .pipe file in a supported model layout.');
  const result = new Map<string, string>();
  const seen = new Set<string>();
  clean.forEach(([original, name]) => {
    if (!name.startsWith(root)) return;
    const relative = name.slice(root.length);
    if (!relative) return;
    const key = relative.toLowerCase();
    if (seen.has(key)) throw new Error(`Duplicate model archive path: ${relative}`);
    seen.add(key);
    result.set(original, relative);
  });
  result.forEach((relative) => {
    const parts = relative.split('/');
    parts.pop();
    while (parts.length) {
      if (seen.has(parts.join('/').toLowerCase())) throw new Error(`Conflicting model archive path: ${relative}`);
      parts.pop();
    }
  });
  return result;
}

async function openArchive(filename: string): Promise<yauzl.ZipFile> {
  return new Promise((resolve, reject) => {
    yauzl.open(filename, { lazyEntries: true, autoClose: false }, (error, zip) => {
      if (error || !zip) reject(error || new Error('Unable to open ZIP')); else resolve(zip);
    });
  });
}

async function readEntries(zip: yauzl.ZipFile): Promise<yauzl.Entry[]> {
  return new Promise((resolve, reject) => {
    const entries: yauzl.Entry[] = [];
    let total = 0;
    zip.on('error', reject);
    zip.on('entry', (entry: yauzl.Entry) => {
      total += entry.uncompressedSize;
      if (entries.length >= MAX_FILES || total > MAX_BYTES) {
        reject(new Error('Model archive exceeds the file count or uncompressed size limit.'));
        return;
      }
      // eslint-disable-next-line no-bitwise
      const type = (entry.externalFileAttributes >>> 16) & 0o170000;
      if (![0, 0o100000, 0o040000].includes(type)) {
        reject(new Error('Model archives cannot contain symbolic links or special files.'));
        return;
      }
      entries.push(entry);
      zip.readEntry();
    });
    zip.once('end', () => resolve(entries));
    zip.readEntry();
  });
}

export async function importModelPack(settings: Settings, filename: string): Promise<void> {
  if (settings.readonlyMode) throw new Error('Model import is disabled in read-only mode.');
  const root = path.join(settings.dataPath, PipelinesFolderName);
  await fs.ensureDir(root);
  const staging = await fs.mkdtemp(path.join(root, '.import-'));
  let zip: yauzl.ZipFile | undefined;
  try {
    zip = await openArchive(filename);
    const entries = (await readEntries(zip)).filter((entry) => !entry.fileName.endsWith('/'));
    const mapping = modelPackPaths(entries.map((entry) => entry.fileName));
    for (const entry of entries) {
      const relative = mapping.get(entry.fileName);
      if (relative) {
        const target = path.join(staging, relative);
        await fs.ensureDir(path.dirname(target));
        const source = await new Promise<NodeJS.ReadableStream>((resolve, reject) => {
          zip!.openReadStream(entry, (error, stream) => {
            if (error || !stream) reject(error || new Error('Unable to read ZIP entry')); else resolve(stream);
          });
        });
        await streamPipeline(source, fs.createWriteStream(target, { flags: 'wx' }));
      }
    }
    const base = path.basename(filename, path.extname(filename)).replace(/[^\p{L}\p{N}_ .-]/gu, '_').replace(/^[ .]+|[ .]+$/g, '') || 'Imported model';
    let name = base;
    let suffix = 2;
    while (await fs.pathExists(path.join(root, name))) {
      name = `${base} (${suffix})`;
      suffix += 1;
    }
    await fs.move(staging, path.join(root, name), { overwrite: false });
  } finally {
    zip?.close();
    await fs.remove(staging);
  }
}

/**
 * Unpack an archive as it is, every entry under destination. Used for the
 * pack `viame train` writes, whose layout is already the pipeline folder's.
 */
export async function extractModelPackTo(filename: string, destination: string): Promise<string[]> {
  let zip: yauzl.ZipFile | undefined;
  const written: string[] = [];
  try {
    zip = await openArchive(filename);
    const entries = (await readEntries(zip)).filter((entry) => !entry.fileName.endsWith('/'));
    for (const entry of entries) {
      const relative = entry.fileName.replace(/\\/g, '/');
      const parts = relative.split('/');
      if (relative.startsWith('/') || parts.some((p) => p === '..' || p === '' || p.includes(':'))) {
        throw new Error(`Unsafe model archive path: ${entry.fileName}`);
      }
      const target = path.join(destination, ...parts);
      await fs.ensureDir(path.dirname(target));
      const source = await new Promise<NodeJS.ReadableStream>((resolve, reject) => {
        zip!.openReadStream(entry, (error, stream) => {
          if (error || !stream) reject(error || new Error('Unable to read ZIP entry')); else resolve(stream);
        });
      });
      await streamPipeline(source, fs.createWriteStream(target));
      written.push(relative);
    }
  } finally {
    zip?.close();
  }
  return written;
}

export async function exportModelPack(settings: Settings, model: Pipe, destination: string): Promise<void> {
  const root = await fs.realpath(path.join(settings.dataPath, PipelinesFolderName));
  const folder = await fs.realpath(path.dirname(model.pipe));
  if (model.type !== 'trained' || path.dirname(folder) !== root) throw new Error('Select a trained model pack to export.');
  const output = path.resolve(destination);
  const relativeOutput = path.relative(folder, output);
  if (!relativeOutput.startsWith(`..${path.sep}`) && !path.isAbsolute(relativeOutput)) throw new Error('Save the ZIP outside the model pack.');
  // Enumerate first to reject links, rather than exporting files outside the pack.
  const files: string[] = [];
  async function collect(directory: string) {
    for (const name of await fs.readdir(directory)) {
      const file = path.join(directory, name);
      const info = await fs.lstat(file);
      if (info.isSymbolicLink()) throw new Error('Model packs cannot contain symbolic links.');
      if (info.isDirectory()) await collect(file);
      else if (info.isFile()) files.push(file);
      else throw new Error('Model packs cannot contain special files.');
    }
  }
  await collect(folder);
  const temp = await fs.mkdtemp(path.join(path.dirname(output), '.dive-export-'));
  try {
    const archivePath = path.join(temp, 'model.zip');
    await new Promise<void>((resolve, reject) => {
      const stream = fs.createWriteStream(archivePath);
      const archive = archiver('zip', { zlib: { level: 1 } });
      const fail = (error: Error) => { archive.abort(); stream.destroy(); reject(error); };
      stream.on('error', fail);
      archive.on('error', fail);
      archive.on('warning', fail);
      stream.on('close', resolve);
      archive.pipe(stream);
      files.forEach((file) => archive.file(file, { name: path.relative(folder, file).split(path.sep).join('/') }));
      archive.finalize().catch(fail);
    });
    await fs.move(archivePath, output, { overwrite: true });
  } finally {
    await fs.remove(temp);
  }
}
